drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,517 @@
+
+/* period_cycles is not used, set frequency on menuconfig instead. */
+static int pwm_led_esp32_pin_set_cycles(struct device *dev,
+					u32_t pwm, u32_t period_cycles,
+					u32_t pulse_cycles)
+{
+	ARG_UNUSED(period_cycles);
+
+	int speed_mode;
+	int channel;
+	int timer;
+	int ret;
+
+	const struct pwm_led_esp32_config * const config =
+		(struct pwm_led_esp32_config *) dev->config->config_info;
+
+	channel = pwm_led_esp32_get_gpio_config(pwm, config->ch_cfg);
+	if (channel < 0) {
+		return -EINVAL;
+	}
+	speed_mode = channel < 8 ? PWM_LED_ESP32_HIGH_SPEED :
+				   PWM_LED_ESP32_LOW_SPEED;
+
+	timer = config->ch_cfg[channel].timer;
+	/* Now we know which speed_mode and timer is set, then we will convert
+	 * the channel number from (0 - 15) to (0 - 7).
+	 */
+	channel %= 8;
+
+	/* Enable peripheral */
+	set_mask32(DPORT_LEDC_CLK_EN, DPORT_PERIP_CLK_EN_REG);
+	clear_mask32(DPORT_LEDC_RST, DPORT_PERIP_RST_EN_REG);
+
+
+	/* Set timer */
+	ret = pwm_led_esp32_timer_set(speed_mode, timer,
+			config->timer_cfg[speed_mode][timer].bit_num,
+			config->timer_cfg[speed_mode][timer].freq);
+	if (ret < 0) {
+		return ret;
+	}
+
+	/* Set channel */
+	ret = pwm_led_esp32_channel_set(pwm, speed_mode, channel, 0, timer);
+	if (ret < 0) {
+		return ret;
+	}
+	pwm_led_esp32_duty_set(speed_mode, channel, pulse_cycles);
+	pwm_led_esp32_update_duty(speed_mode, channel);
+
+	return ret;
+}
+
+static int pwm_led_esp32_get_cycles_per_sec(struct device *dev, u32_t pwm,
+					    u64_t *cycles)
+{
+	const struct pwm_led_esp32_config *config;
+	int channel;
+	int timer;
+	int speed_mode;
+
+	config = (struct pwm_led_esp32_config *) dev->config->config_info;
+
+	channel = pwm_led_esp32_get_gpio_config(pwm, config->ch_cfg);
+	if (channel < 0) {
+		return -EINVAL;
+	}
+	speed_mode = channel < 8 ? PWM_LED_ESP32_HIGH_SPEED :
+				   PWM_LED_ESP32_LOW_SPEED;
+
+	timer = config->ch_cfg[channel].timer;
+
+	*cycles = config->timer_cfg[speed_mode][timer].freq;
+
+	return 0;
+}
+
+static const struct pwm_driver_api pwm_led_esp32_api = {
+	.pin_set = pwm_led_esp32_pin_set_cycles,
+	.get_cycles_per_sec = pwm_led_esp32_get_cycles_per_sec,
+};
+
+int pwm_led_esp32_init(struct device *dev)
+{
+	return 0;
+}
+
+/* Initialization for PWM_LED_ESP32 */
+#include <device.h>
+#include <init.h>
+
+DEVICE_DECLARE(pwm_led_esp32_0);
+
+#define CH_HS_TIMER(i) ((CONFIG_PWM_LED_ESP32_HS_CH ## i ## _TIMER) & 0x2)
+#define CH_HS_GPIO(i) ((CONFIG_PWM_LED_ESP32_HS_CH ## i ## _GPIO) & 0xfff)
+#define CH_LS_TIMER(i) ((CONFIG_PWM_LED_ESP32_LS_CH ## i ## _TIMER) & 0x2)
+#define CH_LS_GPIO(i) ((CONFIG_PWM_LED_ESP32_LS_CH ## i ## _GPIO) & 0xfff)
+
+#define TIMER_HS_FREQ(i) (CONFIG_PWM_LED_ESP32_HS_TIMER ## i ## _FREQ)
+#define TIMER_LS_FREQ(i) (CONFIG_PWM_LED_ESP32_LS_TIMER ## i ## _FREQ)
+#define TIMER_HS_BIT_NUM(i) (CONFIG_PWM_LED_ESP32_HS_TIMER ## i ## _BIT_NUM)
+#define TIMER_LS_BIT_NUM(i) (CONFIG_PWM_LED_ESP32_LS_TIMER ## i ## _BIT_NUM)
+
+static struct pwm_led_esp32_config pwm_led_esp32_config = {
```

Comment:
Can this be `const`? Saves some bytes in RAM. (Not going to make a difference right now, since Zephyr on ESP32 is running from RAM all the time, but the plan is to have most of it in ROM at some point in the future.)